### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ArchitWagle**
Basic structure in place, all tests except gps working

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,396 @@
...
 
-This repository showcases the BrowserStack integration features with the TestCafe using the BrowserStack sample demo canonical app. More details: Customer Demo Repositories
+## Introduction
+
+~intro of framework~. In this document we will run sample tests designed for [browserstack-canonical-app](https://bstackdemo.com/) in this repository on different infrastructures in different configurations. The browserstack-canonical-app is a demo e-commerce application to demonstrate real-world usage of BrowserStack testing methods, patterns, and workflows. It is written using Next/React. The app is bundled with offers data, orders data and products data that contains everything you need to start using the app and run tests out-of-the-box.
```
Comment (mudassarsyed) [inline README.md]: Can you write an Introduction to the repo here

File: README.md [hunk 2]
```diff
@@ -1,3 +1,396 @@
...
+
+## Setup
+
+- Clone this repository
+
+  `git clone https://github.com/browserstack/browserstack-examples-~framework1~-~framework2~`
+
+- Install the following dependencies
```
Comment (mudassarsyed) [inline README.md]: Here you will mention all the language installation of test cafe

File: README.md [hunk 3]
```diff
@@ -1,3 +1,396 @@
...
-This repository showcases the BrowserStack integration features with the TestCafe using the BrowserStack sample demo canonical app. More details: Customer Demo Repositories
+## Introduction
+
+~intro of framework~. In this document we will run sample tests designed for [browserstack-canonical-app](https://bstackdemo.com/) in this repository on different infrastructures in different configurations. The browserstack-canonical-app is a demo e-commerce application to demonstrate real-world usage of BrowserStack testing methods, patterns, and workflows. It is written using Next/React. The app is bundled with offers data, orders data and products data that contains everything you need to start using the app and run tests out-of-the-box.
+
+## Setup
+
+- Clone this repository
+
+  `git clone https://github.com/browserstack/browserstack-examples-~framework1~-~framework2~`
```
Comment (mudassarsyed) [inline README.md]: Replace the repo link here

File: package.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+
+  "scripts": {
+
+    "go": "sh resources/conf/caps/linux.sh"
```
Comment (mudassarsyed) [inline package.json]: Are you trying to run all the profiles one after the other here

File: linux.sh [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+    
+    # set the common env variables from the `common_env` function defined above
+    # these common environment variables are necesaary for the browerstack plugin
+    common_env
+
+    #start local 
+    start_local
+
+    if   [ $suite == "single" ]; then
```
Comment (mudassarsyed) [inline linux.sh]: can we call it $profile

File: linux.sh [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        run_single_test 
+
+    elif [ $suite == "parallel" ]; then
+        run_parallel_Nt_1b
+
```
Comment (mudassarsyed) [inline linux.sh]: Can we use the profile names mentioned on the confluence doc https://browserstack.atlassian.net/wiki/spaces/CE/pages/2091057195/Customer+Demo+Repositories

File: src/test/suites/e2e/test8.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import HomePage from '../../../app/pages/HomePage';
```
Comment (mudassarsyed) [inline src/test/suites/e2e/test8.js]: I can see that you have used the test case number with test keyword, but can we have an appropriate test case name

File: src/test/utils/roles/roles.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    const loginButton           = Selector('#login-btn');
+
+    await t
+        .typeText(userNameInput,userName)
+        .pressKey('enter')
+        .typeText(passwordInput,password)
+        .pressKey('enter')
+        .click(loginButton)
+        
+});
```
Comment (mudassarsyed) [inline src/test/utils/roles/roles.js]: can we think of modularising this

File: src/test/suites/user/test5.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+test("Click on Sign In button and Login as image_not_loading_user", async (t) => {
+
+    const userName           = "image_not_loading_user";
+    const userNameAfterLogin = Selector('.username');
+    
+    await t
+        .useRole(image_not_loading_user)
+        .expect(userNameAfterLogin.innerText)
+        .eql(userName);
```
Comment (mudassarsyed) [inline src/test/suites/user/test5.js]: We have discussed to fail this test here: https://browserstack.slack.com/archives/G01Q0GYTS75/p1615812356033900

File: linux.sh [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+# read the env type and suite name from commandline arguments
+# the env_type can be "on-prem" or "remote"
+# set on-prem to run suite locally, and remote to run on suite browserstack
+# set suite name, we have defined different test combinations depending on suite name which cover 
+# a broad range of usecases
+env_type=$1
+suite=$2
+
```
Comment (mudassarsyed) [inline linux.sh]: Would windows users be able to run the tests with sh?

File: README.md [hunk 11]
```diff
@@ -1,3 +1,396 @@
...
+
+- [On-premise/self-hosted](#on-premise-self-hosted)
+- [Docker](#docker)
+- [BrowserStack](#browserstack)
+
+## Generating Reports
+
+Generate Report using the following command: `~generate-reports-command~`
+
+## Notes
```
Comment (mudassarsyed) [inline README.md]: Can you populate relavant data accross the readme

File: package.json (hunk 4 above)
Comment (mudassarsyed) [reply package.json]: For Simplicity , do you think adding different commands for different profiles would help, example script section: https://github.com/rquellh/testcafe-cucumber/blob/master/package.json

**Review by mudassarsyed — CHANGES_REQUESTED** [review]
Lets brainstorm to make this more better

File: package.json (hunk 4 above)
Comment (ArchitWagle) [reply package.json]: This command is not necessary any more, will remove it. Now the launching is done through the shell script not the package.json

File: linux.sh (hunk 5 above)
Comment (ArchitWagle) [reply linux.sh]: done

File: src/test/suites/e2e/test8.js (hunk 7 above)
Comment (ArchitWagle) [reply src/test/suites/e2e/test8.js]: done

File: src/test/suites/user/test5.js (hunk 9 above)
Comment (ArchitWagle) [reply src/test/suites/user/test5.js]: done

File: linux.sh (hunk 6 above)
Comment (ArchitWagle) [reply linux.sh]: done

File: linux.sh (hunk 10 above)
Comment (ArchitWagle) [reply linux.sh]: yes, they would simply need to install git [stack overflow link](https://stackoverflow.com/questions/26522789/how-to-run-sh-on-windows-command-prompt)

later we can also at a bat file

File: src/test/utils/roles/roles.js (hunk 8 above)
Comment (ArchitWagle) [reply src/test/utils/roles/roles.js]: done

File: linux.sh [hunk 12]
```diff
@@ -21,11 +21,12 @@ common_env(){
...
     # browserstack credentials
-    export BROWSERSTACK_USERNAME=""
-    export BROWSERSTACK_ACCESS_KEY=""
+    #export BROWSERSTACK_USERNAME=""
+    #export BROWSERSTACK_ACCESS_KEY=""
 
     # set the build name, a build is a logical grouping of tests on the automate dashboard
-    export BROWSERSTACK_BUILD_ID="test-cafe"
+    time_stamp=$(date +"%Y-%m-%d %H:%M:%S")
+    export BROWSERSTACK_BUILD_ID="test-cafe-{$(date +"%Y-%m-%d %H:%M:%S")}"
```
Comment (mudassarsyed) [inline linux.sh]: Use EPOC time with buildname

**Review by mudassarsyed — APPROVED** [review]

File: linux.sh (hunk 10 above)
Comment (mudassarsyed) [reply linux.sh]: Since you are working on the readme can u add in the readme

Comment (mudassarsyed) [reply linux.sh]: Can yo add a comment with profile name for every profile in the runner.sh file

File: Jenkinsfile [hunk 13]
```diff
@@ -0,0 +1,89 @@
+import org.jenkinsci.plugins.pipeline.modeldefinition.Utils
```
Comment (samiran-bs) [inline Jenkinsfile]: let's remove this file until it is a functional accurate Jenkinsfile without placeholders.

File: README.md [hunk 14]
```diff
@@ -1,3 +1,357 @@
...
+## Repository setup
+
+- Clone the repository
+
+- Ensure you have the following dependencies installed on the machine
+
+  - npm >= 8
+  - shell
+
+To install bash on windows, just install GIT. During installation of GIT, add GIT Bash to windows context menu by selecting its option. After installation right click in your folder select `GIT Bash Here`. More details over [here](https://stackoverflow.com/questions/26522789/how-to-run-sh-on-windows-command-prompt)
```
Comment (samiran-bs) [inline README.md:25]: this has my conditional approval for the R1 release. would require a JS runner update in the next iteration.

File: README.md [hunk 15]
```diff
@@ -1,3 +1,357 @@
...
+
+npm:
+
+```sh
+npm install
+```
+
+## About the tests in this repository
+
+This repository contains the following #{ Selenium test / Cypress / Puppeteer / Other } tests:
```
Comment (samiran-bs) [inline README.md]: mention TestCafe

File: README.md [hunk 16]
```diff
@@ -1,3 +1,357 @@
...
+- Clone the repository
+
+- Ensure you have the following dependencies installed on the machine
+
+  - npm >= 8
+  - shell
+
+To install bash on windows, just install GIT. During installation of GIT, add GIT Bash to windows context menu by selecting its option. After installation right click in your folder select `GIT Bash Here`. More details over [here](https://stackoverflow.com/questions/26522789/how-to-run-sh-on-windows-command-prompt)
+
+To install all the node modules dependencies, run the following command in the repository root
```
Comment (samiran-bs) [inline README.md]: repository root directory

File: README.md [hunk 17]
```diff
@@ -1,3 +1,357 @@
...
+
+- How to run the test?
+
+  - To run the default test scenario (e.g. End to End Scenario) on your own machine, use the following command:
+
+  ```sh
+  sh runner.sh bstack single
+  ```
+
+  To run a specific test scenario, use the following command with the additional 'relative-test-path' argument:
```
Comment (samiran-bs) [inline README.md]: use "< relative-test-path >" throughout.

File: README.md [hunk 18]
```diff
@@ -1,3 +1,357 @@
...
+
+- Output
+
+  This run profile executes the entire test suite in parallel on a single BrowserStack browser. Please refer to your [BrowserStack dashboard](https://automate.browserstack.com/) for test results.
+
+  - Note: By default, this execution would run maximum 5 test threads in parallel on BrowserStack. Refer to the section ["Configuring the maximum parallel test threads for this repository"](#Configuring-the-maximum-parallel-test-threads-for-this-repository) for updating the parallel thread count based on your requirements.
+
+### Run the entire test suite in parallel on multiple BrowserStack browsers
+
+In this section, we will run the tests in parallel on multiple browsers on Browserstack. Refer to the `run_parallel_1t_Nb` function in `runner.sh` file to change test capabilities for this configuration.
```
Comment (samiran-bs) [inline README.md:269]: can we comment above this line in the code "runner.sh"? w.r.t what exactly is 1t_NB

OR rename the test method: run_parallel_one_test_n_browsers

File: README.md [hunk 19]
```diff
@@ -1,3 +1,357 @@
...
+
+  To run the entire test suite in parallel on multiple BrowserStack browsers, use the following command:
+
+  ```sh
+  sh runner.sh bstack local-parallel-browsers
+  ```
+
+## Generating Reports
+
+- Generate Report using the following flag: `-r html:reports/report.html` in the testcafe command, the node package `testcafe-reporter-html` is needed. Other [test reporters](https://devexpress.github.io/testcafe/documentation/guides/concepts/reporters.html) can also be used
```
Comment (samiran-bs) [inline README.md]: Note: The node package `testcafe-reporter-html` is used in the package.json for supporting this functionality.

Comment (samiran-bs) [reply README.md]: Other test reporters, as outlined in the TestCafe documentation [here](https://devexpress.github.io/testcafe/documentation/guides/concepts/reporters.html) can also be used

File: package-lock.json [hunk 20]
```diff
@@ -0,0 +1,12247 @@
+{
```
Comment (samiran-bs) [inline package-lock.json]: this file should never be checked in. It gets auto generated on npm install

File: reports/report.html [hunk 21]
```diff
@@ -0,0 +1,136 @@
+<!DOCTYPE html>
```
Comment (samiran-bs) [inline reports/report.html]: file should not be checked in

File: runner.sh [hunk 22]
```diff
@@ -0,0 +1,350 @@
+#!/bin/sh
```
Comment (samiran-bs) [inline runner.sh:1]: Please do not check in the binary files above.
resources/local/BrowserStackLocal.*
Have the details mentioned in the doc w.r.t what needs to be installed OR better follow below:
- have the BrowserStackLocal package in the package.json
- After npm install is done, which downloads the package here: ~/.browserstack/ folder ... use it from there.
- There is a similar path in Windows too mostly <USER_HOME_DIR>/.browserstack/. Check once

File: runner.sh [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+
+# run one test on multiple browsers
+run_parallel_1t_Nb(){
+
+    test_base_path="src/test/suites"
+
+    for test_path in $(find $test_base_path -type f -print)
+    do
+    
+        $testcafe "@browserStack/browserstack:firefox@74.0:OS X High Sierra"  $test_path --test-scheduling --reporter spec &
```
Comment (samiran-bs) [inline runner.sh:105]: can these values also come from an array that the user can populate at one go?

@browserStack/browserstack:firefox@74.0:OS X High Sierra
@browserStack/browserstack:chrome@80.0:OS X High Sierra
@browserStack/browserstack:firefox@75.0:Windows 8.1
@browserStack/browserstack:Samsung Galaxy S20@10.0
@browserStack/browserstack:iPhone XS@13.0

File: runner.sh [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+        
+        else
+        # notice the & at the end. This means that the next test would be run in parallel with this test
+        # this command is executed when i=0,1,...,max_parallel-2
+            $testcafe "$browser"  "$test_path" --test-scheduling --skip-js-errors &
+        fi
+        # increment i
+        i=$((i+1))
+    done
+    wait
```
Comment (samiran-bs) [inline runner.sh:153]: what does this wait do?

File: runner.sh [hunk 25]
```diff
@@ -0,0 +1,350 @@
...
+        # covers both mobile and desktop browsers
+        run_parallel_1t_Nb
+
+    # all the below profiles are similar to the above profiles except, 
+    # the test base url is set to a localhost address. 
+    # overwrites the base_url set in function `common_env` since we are trying out local-testing
+
+
+    elif [ $profile == "local" ]; then
+        export TEST_BASE_URL="http://localhost:3000/"
```
Comment (samiran-bs) [inline runner.sh]: can all exports be defined together initially in the file?
use an if else condition for bstackdemo.com vs localhost.

File: src/app/pages/Checkout.js [hunk 26]
```diff
@@ -0,0 +1,32 @@
+import { Selector, t } from 'testcafe';
```
Comment (samiran-bs) [inline src/app/pages/Checkout.js:1]: any reason Bag.js file is empty?

File: src/app/pages/HomePage.js [hunk 27]
```diff
@@ -0,0 +1,47 @@
+import { Selector, t } from 'testcafe';
```
Comment (samiran-bs) [inline src/app/pages/HomePage.js:1]: any reason Favourites.js file is empty?

File: src/app/pages/Orders.js [hunk 28]
```diff
@@ -0,0 +1,19 @@
+import { Selector, t } from 'testcafe';
```
Comment (samiran-bs) [inline src/app/pages/Orders.js:1]: any reason Offers.js file is empty?

File: src/test/suites/login/LockedUserTest.js [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+    const password             = "testingisfun99";
+    const signInButton         = Selector('#signin');
+    const userNameInput        = Selector('#username input');
+    const passwordInput        = Selector('#password input');
+    const loginButton          = Selector('#login-btn');
+    const errorMessage         = Selector('.api-error');
+    const expectedErrorMessage = "Your account has been locked.";
+     
+    // test
+    // Note: could have used `roles` here for login too.
```
Comment (samiran-bs) [inline src/test/suites/login/LockedUserTest.js]: lets do the needful and cleanup the comments like these which are not user facing.

File: src/test/suites/offers/GPSLocationTest.js [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+import { userRolesDict } from '../../utils/roles/roles'
+
+
+fixture("offers")
+    .page(process.env.TEST_BASE_URL);
+
+
+// overwrite the getCurrentPosition func to mock Mumbai's geolocation
+const mockGeoLocationForMumbai = ClientFunction(() => {
+    window.navigator.geolocation.getCurrentPosition = function(cb){cb({ coords: {accuracy: 20,altitude: null,altitudeAccuracy: null,heading: null,latitude: 1,longitude: 103,speed: null}}); }
```
Comment (samiran-bs) [inline src/test/suites/offers/GPSLocationTest.js:11]: run all files through a formatter.

File: src/test/suites/product/ProductFiltersTest.js [hunk 31]
```diff
@@ -0,0 +1,44 @@
+import { Selector } from 'testcafe'
+
+fixture("product")
+    .page(process.env.TEST_BASE_URL);
+
+test("Apply Apple and Samsung Filters", async (t) => {
+
+    // As of now this test is set to filter by apple or samsung
+    // to filter for any other device manufaurers modify the regex for the `checkBoxes` selector
+    await t.wait(2000);
```
Comment (samiran-bs) [inline src/test/suites/product/ProductFiltersTest.js]: please identify a better way to do away with the hardcoded wait.
can there be an element which could be waited upon?

File: src/test/suites/product/ProductFiltersTest.js [hunk 32]
```diff
@@ -0,0 +1,44 @@
+import { Selector } from 'testcafe'
+
+fixture("product")
+    .page(process.env.TEST_BASE_URL);
+
+test("Apply Apple and Samsung Filters", async (t) => {
+
+    // As of now this test is set to filter by apple or samsung
+    // to filter for any other device manufaurers modify the regex for the `checkBoxes` selector
```
Comment (samiran-bs) [inline src/test/suites/product/ProductFiltersTest.js]: if we are adding comments, please have a spell check done.

e.g. manufaurers

**Review by samiran-bs — CHANGES_REQUESTED** [review]

File: runner.sh (hunk 22 above)
Comment (ArchitWagle) [reply runner.sh:1]: As of now due to time constraint with trying out on windows did it using the first method, 
will implement in the second soon

**Review by mudassarsyed — APPROVED** [review]

**Pull request merged**